Ansible Module Guidelines

docschick opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Docs Pull Request
##### ANSIBLE VERSION

```
ansible --devel
```
##### SUMMARY

Google doc collaboration from various Ansible team members merged into current module development doc.

Purpose of this work is to add clarity and information regarding module development guidelines.

commit message: "start of merge from google doc and team comments, module guidelines as focus, added a link anchor as needed"

Commit:
start of merge from google doc and team comments, module guidelines as focus, added a link anchor as needed

Commit:
adding links and anchors for links, referencing check mode

bcoca commented:
I'm tempted to say that these should just be part of the checklist instead of their own section

juliedavila commented on docsite/rst/developing_modules.rst:
```diff
@@ -23,6 +23,51 @@ repo for more established and widely used modules.  "Extras" modules may be prom
+- Always return useful data, even when there is no change.
+- Be consistent about returns (some modules are too random), unless it is detrimental to the state/action.
+- Make returns reusable--most of the time you don't want to read it, but you do want to process it and re-purpose it.
+- return diff is in diff mode?
```
Not sure what this line is about

docschick replied:
The Google doc has this comment in regard to "return diff is in diff mode":

Timothy Appnel/ 10:47 AM Mar 17 -- Thinking we'd want this. If I'm correct we should provide a bit of guidance on what to return here also.

tima replied:
I also think that should have read "return diff IF in diff mode."

Commit:
trying the new content in the checklist instead of new sections

per bcoca's comment, thought I would move the new content to the checklist
to see if you guys think this is a better treatment.

docschick commented on docsite/rst/developing_modules.rst:
```diff
@@ -494,7 +497,24 @@ Module checklist
+    * Always return useful data, even when there is no change.
+    * Be consistent about returns (some modules are too random), unless it is detrimental to the state/action.
+    * Make returns reusable--most of the time you don't want to read it, but you do want to process it and re-purpose it.
+    * Return diff if in diff mode
```
Anything else to add or say about "Return diff if in diff mode"? Wanted to make sure this was addressed if there were still questions around it.

juliedavila replied:
Nah I think it's good now. Previously the grammar was confusing me as it looked like a question. But looks good now

bcoca replied:
check and diff mode are 'brownie points' in my book, i would not require all modules to do so (does not make sense in some) but would encourage them when appropriate.

gundalow commented on docsite/rst/developing_modules.rst:
```diff
@@ -478,9 +478,12 @@ gives them slightly higher development priority (though they'll work in exactly
 
+The following  checklist items are important guidelines for people who want to contribute to the development of modules to Ansible on GitHub. Please read the guidelines before you submit your PR/proposal.
+
 * The shebang should always be #!/usr/bin/python, this allows ansible_python_interpreter to work
```
`#!/usr/bin/python`

gundalow commented on docsite/rst/developing_modules.rst:
```diff
@@ -478,9 +478,12 @@ gives them slightly higher development priority (though they'll work in exactly
 * Modules must be written to support Python 2.4. If this is not possible, required minimum python version and rationale should be explained in the requirements section in DOCUMENTATION.
 * Documentation: Make sure it exists
+    * Module documentation should briefly and accurately define what each module and option does, and how it works with others in the underlying system. Documentation should be written for broad audience--readable both by experts and non-experts. This documentation is not meant to teach a total novice, but it also should not be reserved for the Illuminati (hard balance).
     * `required` should always be present, be it true or false
```
`true` or `false`

gundalow commented:
Hi,

+1 for improving documentation

Some other feedback based on some recent new modules I've reviewed

1) I've seen a few modules just having the `HAS_LIB=True` section without actually checking it in main

```
if not HAS_LIb:
  module.fail_json(msg='library FOO is required.')
```

2) Descriptions need Capital letter & full stop. This make everything look a lot more consistent.

3) If an optional parameter is sometimes required this need to be reflected in the documentation, e.g.
"Required when C(state=present)."

Note: required MUST be exactly "true" or "false". "false, required if state=true" actually resolved to "true" when the documentation is generated

4) There isn't any mention of `extends_documentation_fragment`
For example, and AWS module should have

```
 extends_documentation_fragment:
     - aws
     - ec2
```

5) There has been talk that examples in new modules should use `key: value`, rather than `key=value`, though I'm not sure if that's been signed off yet. @abadger @bcoca @gregdek Need guidence from you I guess, happy to create a ticket/proposal if you think that's a better place for discussion?

Similarly, should state be a required parameter (for new modules) since we tell people to always set it in their playbooks. 

6) From ansible-meetings there was a suggestion to add a clear link to the separate page about developing Cloud/AWS modules. Given from a quick Google I can't find the page, that would be worth while.

gregdek commented:
I agree about trying to incorporate module-dependent guidelines in a more centralized place. Currently we have additional Amazon module guidelines here:

https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/amazon/GUIDELINES.md

But I'm concerned that people miss these too easily. We could, at the very least, note that additional module guidelines may exist for certain families of modules, with pointers to where they might be.

Commit:
attempting to wrangle feedback from git comments.
please edit specific lines when possible to ensure your feedback
is incorporated properly. thanks!

Commit:
minor typo fixes

fixed a few minor typos in the intro for the section
"Deprecating and making module aliases"

Commit:
Merge remote-tracking branch 'ansible/devel' into devel

Commit:
Revert "Merge remote-tracking branch 'ansible/devel' into devel"

This reverts commit 60d7d03c1da6de79824c443dc715e45bf2b5af0e, reversing
changes made to 5ac134f5e0a35be70fe6ca277cbfe3e14062f666.

sivel commented:
Looks like there are a lot more than docs changes in here now.  306 changed files currently.  Might require a rebase instead of a merge.

docschick commented:
@sivel yeah, I did _something_ really wrong when trying to resolve the conflicts for this merge and I've just mucked it up. @jimi-c has been alerted to my snafu and has offered to help in a bit, when he's free.

jimi-c commented:
Resolved and merged manually, so closing this now. Thanks!
